untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+and multiplication
+$$mul = \lambda m\; n. \lambda s\; z. m\;(n\;s)\;z.$$
+The former counts $m$ starting from $n$, and the latter counts $m$ times $n$ starting from $z.$
+
+We can also easily test whether a value corresponds to zero (i.e. $c_0$) or not:
+$$isZero = \lambda n. n\;(\lambda x. fls)\;tru.$$
+
+We run into a bit of an issue when trying to define subtraction or a predecessor function.
+Turns out it's doable but neither pretty nor efficient.
+We'll use pairs of values containing two subsequent naturals.
+That way, we can stagger the counter by one.
+To simplify the notation, we'll introduce two named helper abstractions:
+
+$zz = pair\;c_0\;c_0$\
+$ss = \lambda nn. pair\;(snd\; nn)\;(succ\;(snd\;nn)).$\
+
+The first element of the pair here is the predecessor, and the second is the successor.
+$zz$ encodes the starting point, and $ss$ is the staggered successor.
+
+Now we can define the predecessor function:
+
+$$pred = \lambda n. fst\; (n\; ss\; zz).$$
+
+With this definition, $pred\;c_0 = c_0,$ which isn't ideal.
+However, we don't have an encoding for negative numbers, so we don't have much of a choice.
+If we need to, we can use any other value as the first element of $zz$ to signify failure.
+
+### Recursion
+
+While we're on the topic of signifying failure, there's a curious case of expressions for which beta-reduction never stops.
+
+Such expressions don't have a normal form, and we call those _diverging_.
+
+The simplest diverging expression is the application of the so-called omega combinator to itself:
+
+$$\Omega = \lambda x. x\;x.$$
+
+Using any typical evaluation strategy, one step of beta-reduction of $\Omega\;\Omega$ gives us $\Omega\;\Omega$ again.
+
+The omega combinator is somewhat useless on its own.
+However, we can extend it to be a little more useful.
+In particular, we can encode recursion using the fixed point combinator, also known as $Y$ combinator:
+
+$$Y = \lambda f. (\lambda x. f\;(x\;x))\;(\lambda x. f\;(x\;x)).$$
+
+A shorter equivalent form is sometimes cited:
+
+$$X = \lambda f. \Omega\; (\lambda x. f\;(x\;x)).$$
+
+We should note that it's not the only fixed-point combinator.
+For instance, another famous one, called the Turing's combinator (named after its discoverer), is
+
+$$\Theta = (\lambda x\;y.y\; (x\; x\; y))\; (\lambda x\;y.y\; (x\; x\; y)).$$
+
+The idea here is to give us the ability to express recursion by passing the function as its first argument.
+The omega combinator infinitely replicates itself.
+The $Y$ combinator exploits the same potentially infinitely replicating structure to encode recursion.
+
+Let us see how it works in practice.
+Consider a function that counts down to zero using the $pred$ combinator, defined in the previous section, and returns zero once the argument is zero.
+To define it, we will add a recursive call as its first argument:
+
+$cdown = \lambda f. \lambda n. ifThenElse\;(isZero\;n)\;c_0\;(f\;(pred\;n)).$
```

Comment:
Point at the end is the ending of the sentence? 

Since it renders in a new line, I don't think it is necessary here and elsewhere.

Reply:
Eh, that's actually standard practice for formulae-heavy articles. It's harder to parse without puncutation.

Reply:
Ok, if that's the case, then I have no objections.